docs/adr/0002-database-corruption.md

```diff
@@ -0,0 +1,132 @@
+* Keeping a corrupted database around is almost as bad.  It currently prevents
+  the component from working at all.
+* We don't currently have a good way to distinguish between the two error
+  cases, but this can be improved by reviewing telemetry and sentry data.
+
+## Considered Options
+
+* A) When do we identify corrupted databases?
+  * At migration time only
+  * Whenever we perform a database operation
+* B) How do we identify corrupted databases?
+  * Assume all errors during migration are from corrupt databases
+  * Compare error values against known corruption errors for all operations
+  * Compare error values against known corruption errors when opening the database
+* C) What do we do when we identify corrupted databases?
+  * Delete the database file and recreate the database
+  * Move the database file and recreate the database
+  * Have the component fail
+
+## Decision Outcome
+
+* A: Whenever we perform a database operation
+
+This was chosen because it can be a long time between migrations and we want
+to de able to fix corrupt database during that period.
+
+* B: Compare error values against known corruption errors when opening the database
+
+This was chosen because it had the simplest initial implementation and was
+least likely to misidentify errors.  We can target only errors that clearly
+indicate corruption, for example the `DatabaseCorrupt` error code.  We can also
+only target opening the database, which is the easiest time to fix issues.
+
+* C: Delete the database file and recreate the database
+
+This follows the choice of B.  Since we're being conservative in identifying
+errors, we can delete the database file with relative confidence.
+
+### Implementation notes:
+
+This can be implemented by updating the `Store` code to wrap each database
+operation with a check for errors that we can positively identify as
+corruption-related.  If we see one of those errors, then we can delete the
+database file and start from scratch.
+
+Also, we will start tracking Sentry errors and telemetry to identify errors
+that we know are from SQLite corruption.
+
+# Pros and Cons of the Options
+
+### A: Check for corruption at migration time only
+* Good, because corruption errors are more likely to be noticed at migration time.
+* Good, because it's a natural time to delete the database -- since the user hasn't run any queries yet.
+* Bad, migrations happen infrequently.  Errors that manifest during migrations
+  were probably also causing issues before.
+
+### A: Check for corruption whenever we perform a database operation (Decided)
+* Good, because this gives us the opportunity to fix corruption errors sooner.
+* Bad, because the consumer code probably doesn't expect the database to be
+  deleted and recreated in the middle of a query.  However, this is just an
+  extreme case of normal database behavior -- for example any given row can be
+  deleted during a sync.
```

Comment:
In the case of a corrupted database, I presume the caller was already going to receive *some sort of error* here anyway, so perhaps it's not really that bad for them in practice. We're making a problem that they already had more obvious, but not introducing a new problem.

Reply:
Actually, I'm not sure I fully understand what's being proposed here in the combination of decided options for (A) and (B). The decided option here for (A) says we will check for corruption whenever we perform a database operation, but the decided option for (B) says we will only do it when opening the database.

Are we checking on every operation, or only on every open?

Reply:
That represents the fact that I was going back and forth between the two options, missed some edits, and the doc ended up with some confusion.  I think the decision was to start by only checking at open time, see how that affects the sentry errors/telemetry, and then make a decision about how to tackle the rest.